src/index/StringSortComparator.h

```diff
@@ -0,0 +1,462 @@
+      raise(err);
+      // todo<joka921> : make this customizable for future versions
+      col->setMaxVariable(UCOL_REORDER_CODE_SYMBOL, err);
+      raise(err);
+    }
+  }
+
+  // convert LESS EQUAL GREATER from icu to -1, 0, +1 to make it equivalent to
+  // results from std::strcmp
+  static int compToInd(const UCollationResult res) {
+    switch (res) {
+      case UCOL_LESS:
+        return -1;
+      case UCOL_EQUAL:
+        return 0;
+      case UCOL_GREATER:
+        return 1;
+    }
+    throw std::runtime_error(
+        "Illegal value for UCollationResult. This should never happen!");
+  }
+
+  /// This conversion is needed for "older" versions of ICU, e.g. ICU60 which is
+  /// contained in Ubuntu's LTS repositories
+  static icu::StringPiece toStringPiece(std::string_view s) {
+    return icu::StringPiece(s.data(), s.size());
+  }
+};
+
+/**
+ * \brief This class compares strings according to proper Unicode collation,
+ * e.g. Strings from the text index vocabulary To Compare components of RDFS
+ * triples use the TripleComponentComparator defined below
+ */
+class SimpleStringComparator {
+ public:
+  using Level = LocaleManager::Level;
+  using Transformed_T = LocaleManager::WeightString;
+  SimpleStringComparator(const std::string& lang, const std::string& country,
+                         bool ignorePunctuationAtFirstLevel)
+      : _locManager(lang, country, ignorePunctuationAtFirstLevel) {}
+  SimpleStringComparator() = default;
+
+  /**
+   * \brief Compare two Strings
+   * @return True iff a comes before b
+   */
+  bool operator()(std::string_view a, std::string_view b,
+                  const Level level = Level::IDENTICAL) const {
+    return _locManager.compare(a, b, level) < 0;
+  }
+
+  // ________________________________________________________________________________________________
+  bool operator()(std::string_view a, const LocaleManager::WeightString& b,
+                  const Level level) const {
+    auto aTrans = _locManager.getSortKey(a, level);
+    auto cmp = _locManager.compare(aTrans, b, level);
+    return cmp < 0;
+  }
+
+  /// transform a string s to the collation weight string of the first possible
+  /// string that compares greater to s according to the held locale. Needed for
+  /// Prefix search. Currently only supports the primary collation Level!!!
+  /// <TODO<joka921>: Implement this on every level, either by fixing ICU or by
+  /// hacking the collation strings
+  [[nodiscard]] LocaleManager::WeightString transformToFirstPossibleBiggerValue(
+      std::string_view s) const {
+    auto transformed = _locManager.getSortKey(s, Level::PRIMARY);
+    unsigned char last = transformed.get().back();
+    if (last < std::numeric_limits<unsigned char>::max()) {
+      transformed.get().back() += 1;
+    } else {
+      transformed.get().push_back('\0');
+    }
+    return transformed;
+  }
+
+      [[nodiscard]] const LocaleManager& getLocaleManager() const {
```

Comment:
Formatting seems wrong, maybe we finally need a newer `clang-format` that knows `[[nodiscard]]`

Reply:
Yes, this is even simple with our Ubuntu/Docker setup, see #299

Reply:
Thanks, I merged that